File: tasks/music_player_app/webapp/src/artwork.js
/**
 * Artwork fetcher — searches Discogs, Deezer, iTunes for album art.
 *
 * Session-level dedupe only. Persistent caching lives in the unified
 * `tracks` IDB store, attached to the per-track row by main.js once a
 * lookup succeeds (see updateTrackArtwork in telegram.js).
 */
import { corsFetch } from './cors-proxy.js';

const TIMEOUT = 10000;
const cache = {}; // `${title}|${artist}` -> url (in-memory, session only)

function _artworkCacheKey(title, artist = '') {
    return `${String(title || '').toLowerCase()}|${String(artist || '').toLowerCase()}`;
}

export async function searchArtwork(title, artist = '') {
    const key = _artworkCacheKey(title, artist);
    if (key in cache) return cache[key];

    // iTunes works without CORS proxy, try it first
    let url = await tryItunes(title, artist);
    if (!url) url = await tryDeezer(title, artist);
    if (!url) url = await tryDiscogs(title, artist);

    cache[key] = url;
    return url;
}

export function getArtworkSource(url) {
    if (!url) return null;
    try {
        const host = new URL(url).hostname.replace(/^www\./, '');
        return host;
    } catch {
        return 'internet';
    }
}

function fetchWithTimeout(url, opts = {}) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT);
    return fetch(url, { ...opts, signal: controller.signal }).finally(() => clearTimeout(timer));
}

async function tryDiscogs(title, artist) {
    try {
        const q = artist ? `${artist} ${title}` : title;
        const url = `https://api.discogs.com/database/search?${new URLSearchParams({ q, type: 'release', per_page: '5' })}`;
        const resp = await corsFetch(url);
        if (!resp) return null;
        const data = await resp.json();
        for (const item of data.results || []) {
            if (item.cover_image && !item.cover_image.includes('spacer.gif')) {
                return item.cover_image;
            }
        }
    } catch { /* ignore */ }
    return null;
}

async function tryDeezer(title, artist) {
    try {
        const q = artist ? `${artist} ${title}` : title;
        const url = `https://api.deezer.com/search?${new URLSearchParams({ q, limit: '5' })}`;
        const resp = await corsFetch(url);
        if (!resp) return null;
        const data = await resp.json();
        const results = data.data || [];
        if (!results.length) return null;

        const tl = title.toLowerCase();
        const al = artist.toLowerCase();
        for (const item of results) {
            const it = (item.title || '').toLowerCase();
            const ia = (item.artist?.name || '').toLowerCase();
            if (tl.includes(it) || it.includes(tl)) {
                if (!artist || al.includes(ia) || ia.includes(al)) {
                    return item.album?.cover_big || item.album?.cover_medium;
                }
            }
        }
        return results[0].album?.cover_big || results[0].album?.cover_medium || null;
    } catch { /* ignore */ }
    return null;
}

async function tryItunes(title, artist) {
    try {
        const q = artist ? `${artist} ${title}` : title;
        const resp = await fetchWithTimeout(
            `https://itunes.apple.com/search?${new URLSearchParams({ term: q, media: 'music', limit: '5' })}`
        );
        if (!resp.ok) return null;
        const data = await resp.json();
        const results = data.results || [];
        if (!results.length) return null;
        const artUrl = results[0].artworkUrl100 || '';
        return artUrl ? artUrl.replace('100x100', '600x600') : null;
    } catch { /* ignore */ }
    return null;
}

File: tasks/music_player_app/artwork.py
import os

import aiohttp

TIMEOUT = aiohttp.ClientTimeout(total=10)
HEADERS = {"User-Agent": "TelegramMusicPlayer/1.0"}


class ArtworkFetcher:

    def __init__(self, cache_dir: str):
        self._cache = {}  # (title, artist) -> url
        self._img_cache_dir = os.path.join(cache_dir, "artwork")
        os.makedirs(self._img_cache_dir, exist_ok=True)

    async def search(self, title: str, artist: str = "") -> str | None:
        """Search for album art URL. Returns URL or None."""
        from tasks.music_player_app.lyrics import parse_track_info
        title, artist = parse_track_info(title, artist)

        cache_key = (title.lower().strip(), artist.lower().strip())
        if cache_key in self._cache:
            return self._cache[cache_key]

        url = await self._try_discogs(title, artist)
        if not url:
            url = await self._try_deezer(title, artist)
        if not url:
            url = await self._try_itunes(title, artist)

        self._cache[cache_key] = url
        return url

    async def download(self, url: str, filename: str) -> str | None:
        """Download image to cache dir. Returns local path."""
        path = os.path.join(self._img_cache_dir, filename)
        if os.path.exists(path) and os.path.getsize(path) > 0:
            return path
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(url, timeout=TIMEOUT) as resp:
                    if resp.status != 200:
                        return None
                    data = await resp.read()
                    with open(path, "wb") as f:
                        f.write(data)
                    return path
        except Exception:
            return None

    async def _try_discogs(self, title: str, artist: str) -> str | None:
        """Search Discogs for album art (free, no API key needed for search)."""
        try:
            q = f"{artist} {title}".strip() if artist else title
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    "https://api.discogs.com/database/search",
                    params={"q": q, "type": "release", "per_page": 5},
                    timeout=TIMEOUT,
                    headers={**HEADERS, "Authorization": "Discogs key=DiscogsMusicPlayer, secret=DiscogsMusicPlayer"},
                ) as resp:
                    if resp.status != 200:
                        return None
                    data = await resp.json()
                    results = data.get("results", [])
                    if not results:
                        return None
                    # Pick the first result with a cover image
                    for item in results:
                        cover = item.get("cover_image")
                        if cover and "spacer.gif" not in cover:
                            return cover
                    return None
        except Exception:
            return None

    async def _try_deezer(self, title: str, artist: str) -> str | None:
        """Search Deezer for album art (free, no API key)."""
        try:
            q = f"{artist} {title}".strip() if artist else title
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    "https://api.deezer.com/search",
                    params={"q": q, "limit": 5},
                    timeout=TIMEOUT,
                    headers=HEADERS,
                ) as resp:
                    if resp.status != 200:
                        return None
                    data = await resp.json()
                    results = data.get("data", [])
                    if not results:
                        return None

                    # Pick best match
                    title_l = title.lower()
                    artist_l = artist.lower()
                    for item in results:
                        item_title = (item.get("title") or "").lower()
                        item_artist = (item.get("artist", {}).get("name") or "").lower()
                        if title_l in item_title or item_title in title_l:
                            if not artist or artist_l in item_artist or item_artist in artist_l:
                                album = item.get("album", {})
                                return album.get("cover_big") or album.get("cover_medium")

                    # Fallback: first result
                    album = results[0].get("album", {})
                    return album.get("cover_big") or album.get("cover_medium")
        except Exception:
            return None

    async def _try_itunes(self, title: str, artist: str) -> str | None:
        """Search iTunes for album art (free, no API key)."""
        try:
            q = f"{artist} {title}".strip() if artist else title
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    "https://itunes.apple.com/search",
                    params={"term": q, "media": "music", "limit": 5},
                    timeout=TIMEOUT,
                    headers=HEADERS,
                ) as resp:
                    if resp.status != 200:
                        return None
                    data = await resp.json()
                    results = data.get("results", [])
                    if not results:
                        return None

                    # Use 600x600 art
                    art_url = results[0].get("artworkUrl100", "")
                    if art_url:
                        return art_url.replace("100x100", "600x600")
                    return None
        except Exception:
            return None
